Add functions re passwords

kylecweeks opened the pull request:
Simple functions to randomize string of characters to generate a password of specified length.

kylecweeks committed:
Add functions re passwords

Simple functions to randomize string of characters to generate a password of specified length.

FISHMANPET commented on UMN-Common.psm1:
```diff
@@ -736,3 +736,109 @@ function Write-Log {
+	Begin{}
+	Process{
+		$random = 1..$length | ForEach-Object {Get-Random -Maximum $characters.length}
+		$private:ofs=""
```
looks like this sets the default separator when joining an array of items, I think it'd be better to be more explicit in the places where you're joining rather than defining it here (see later comments)

FISHMANPET commented on UMN-Common.psm1:
```diff
@@ -736,3 +736,109 @@ function Write-Log {
+	Begin{$characterArray = $inputString.ToCharArray()}
+	Process{
+		$scrambledStringArray = $characterArray | Get-Random -Count $characterArray.Length
+		$outputString = -join $scrambledStringArray
```
This could be `$outputString = $scrambledStringArray -join ''`

FISHMANPET commented on UMN-Common.psm1:
```diff
@@ -736,3 +736,109 @@ function Write-Log {
+	}
+}
+
+function Scramble-String{
```
This is a non-approved verb, but for what should be a "private" function it's probably fine

FISHMANPET commented on UMN-Common.psm1:
```diff
@@ -736,3 +736,109 @@ function Write-Log {
+	End{return $outputString}
+}
+
+function new-password{
```
should be New-Password

FISHMANPET commented on UMN-Common.psm1:
```diff
@@ -736,3 +736,109 @@ function Write-Log {
+		.EXAMPLE
+		new-password -passwordLength 15
+		42
+	#>
```
add something as a `.Note` that says where this "algorithm" came from

FISHMANPET commented on UMN-Common.psm1:
```diff
@@ -736,3 +736,109 @@ function Write-Log {
+		$private:ofs=""
+	}
+	End{
+		return [String]$characters[$random]
```
Looks like this uses the $private:ofs as well, this could be replaced with `return $characters[$random] -join ''`

kylecweeks committed:
tweak

drop plural

FISHMANPET commented on UMN-Common.psm1:
```diff
@@ -736,3 +736,109 @@ function Write-Log {
+		[int]$passwordLength = 40
+	)
+	Begin{
+		$length = [System.Math]::Round(($length = $passwordLength/4),0)
```
I think the intent here is to make sure you always get exactly as many characters as requested, even if the number isn't divisible by 4.  I simplified it quite a bit, using [Math]::Truncate instead of doing string operations, and putting it into a much simpler switch statement, taking advantage of the modulus (`%`) operator:
```powershell
[int]$length = [Math]::Truncate($passwordLength / 4)

switch ($passwordLength % 4) {
    0 { $lengths = @($length, $length, $length, $length) }
    1 { $lengths = @($length, $length, $length, ($length + 1)) }
    2 { $lengths = @($length, $length, ($length + 1), ($length + 1)) }
    3 { $lengths = @($length, ($length + 1), ($length + 1), ($length + 1)) }
}
```

Then at the bottom you can use `$lengths[0]` and `$lengths[1]` etc instead of `$length` and `$length2` etc

kylecweeks replied:
Pretty - yeah - a lot cleaner.

FISHMANPET reviewed (changes requested):
A few small changes requested

kylecweeks committed:
Tweaks after review

Added note section for where this came from. Tweaked per suggestions from review.
